File: 112. Path Sum/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int sum) {
        if ( !root ) return 0;
        if ( root->val == sum && root->left == NULL && root->right == NULL ) return 1;
        return hasPathSum(root->left, sum-root->val) || hasPathSum(root->right, sum-root->val);
    }
};

// BFS + queue
class Solution2 {
public:
    bool hasPathSum(TreeNode* root, int sum) {
        if ( !root ) return 0;
        queue<tuple<TreeNode*, int>> q;
        q.push(make_tuple(root, sum-root->val));
        while ( !q.empty() ) {
            tuple<TreeNode*, int> tp = q.front(); q.pop();
            TreeNode * p = get<0>(tp);
            int val = get<1>(tp);
            if ( !(p->left) && !(p->right) && val == 0 ) return 1;
            if ( p->left ) q.push(make_tuple(p->left, val-p->left->val));
            if ( p->right ) q.push(make_tuple(p->right, val-p->right->val));
        }
        return 0;
    }
};


// DFS + stack
class Solution3 {
public:
    bool hasPathSum(TreeNode* root, int sum) {
        if ( !root ) return 0;
        stack<tuple<TreeNode*, int>> s;
        s.push(make_tuple(root, sum-root->val));
        while ( !s.empty() ) {
            tuple<TreeNode*, int> tp = s.top(); s.pop();
            TreeNode * p = get<0>(tp);
            int val = get<1>(tp);
            if ( !(p->left) && !(p->right) && val == 0 ) return 1;
            if ( p->right ) s.push(make_tuple(p->right, val-p->right->val));
            if ( p->left ) s.push(make_tuple(p->left, val-p->left->val));
        }
        return 0;
    }
};

File: 112. Path Sum/Solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def hasPathSum(self, root: TreeNode, sum: int) -> bool:
        if node is None:
            return False
        if node.val == sum and node.left is None and node.right is None:
            return True
        return self.hasPathSum(node.left, sum-node.val) or self.hasPathSum(node.right, sum-node.val)


# if __name__ == '__main__':
#     node = TreeNode(5)
#     n1 = TreeNode(4)
#     n2 = TreeNode(8)
#     node.left, node.right = n1, n2
#     n3 = TreeNode(11)
#     n4 = TreeNode(13)
#     n1.left = n3
#     n5 = TreeNode(4)
#     n2.left, n2.right = n4, n5
#     n6 = TreeNode(7)
#     n7 = TreeNode(2)
#     n8 = TreeNode(1)
#     n3.left, n3.right = n6, n7
#     n5.right = n8
#     s = Solution()
#     res = s.hasPathSum(node, 22)
